app/uk/gov/hmrc/cdsreimbursementclaimfrontend/controllers/claims/CheckClaimantDetailsController.scala

```diff
@@ -173,49 +208,94 @@ object CheckClaimantDetailsController {
-            val phone     = declarant.contactDetails.flatMap(_.telephone)
-            NamePhoneEmail(name, phone.map(PhoneNumber(_)), Some(email))
-        }
+    draftC285Claim.mrnContactDetailsAnswer
+      .map { contact =>
+        NamePhoneEmail(Option(contact.fullName), Option(contact.phoneNumber), Option(contact.emailAddress))
       }
+      .orElse(
+        Applicative[Option]
+          .map2(draftC285Claim.displayDeclaration, draftC285Claim.declarantTypeAnswer) { (declaration, declarantType) =>
+            declarantType match {
+              case DeclarantTypeAnswer.Importer | DeclarantTypeAnswer.AssociatedWithImporterCompany =>
+                val consignee = declaration.displayResponseDetail.consigneeDetails
+                val name      = consignee.flatMap(_.contactDetails).flatMap(_.contactName)
+                val phone     = consignee.flatMap(_.contactDetails).flatMap(_.telephone)
+                NamePhoneEmail(name, phone.map(PhoneNumber(_)), Some(email))
+              case DeclarantTypeAnswer.AssociatedWithRepresentativeCompany                          =>
+                val declarant = declaration.displayResponseDetail.declarantDetails
+                val name      = declarant.contactDetails.flatMap(_.contactName)
+                val phone     = declarant.contactDetails.flatMap(_.telephone)
+                NamePhoneEmail(name, phone.map(PhoneNumber(_)), Some(email))
+            }
+          }
+      )
       .getOrElse(NamePhoneEmail(None, None, None))
   }
 
   def extractContactAddress(fillingOutClaim: FillingOutClaim): Option[NonUkAddress] = {
     val draftC285Claim = fillingOutClaim.draftClaim.fold(identity)
-    Applicative[Option]
-      .map2(draftC285Claim.displayDeclaration, draftC285Claim.declarantTypeAnswer) { (declaration, declarantType) =>
-        val contactDetails = declarantType match {
-          case DeclarantTypeAnswer.Importer | DeclarantTypeAnswer.AssociatedWithImporterCompany =>
-            declaration.displayResponseDetail.consigneeDetails.flatMap(_.contactDetails)
-          case DeclarantTypeAnswer.AssociatedWithRepresentativeCompany                          =>
-            declaration.displayResponseDetail.declarantDetails.contactDetails
-        }
-        NonUkAddress(
-          contactDetails.flatMap(_.addressLine1).getOrElse(""),
-          contactDetails.flatMap(_.addressLine2),
-          None,
-          contactDetails.flatMap(_.addressLine3).getOrElse(""),
-          contactDetails.flatMap(_.postalCode).getOrElse(""),
-          contactDetails.flatMap(_.countryCode).map(Country(_)).getOrElse(Country.uk)
+    draftC285Claim.mrnContactAddressAnswer
+      .orElse(
+        Applicative[Option]
+          .map2(draftC285Claim.displayDeclaration, draftC285Claim.declarantTypeAnswer) { (declaration, declarantType) =>
+            (declarantType match {
+              case DeclarantTypeAnswer.Importer | DeclarantTypeAnswer.AssociatedWithImporterCompany =>
```

Comment:
This case pattern matches been rep-eaten all over. Presumably I need to repeat them when I need to update address. 
Would it be possible to make this check common?

Reply:
I was thinking about that, but then I would be returning a huge tuple.